torch/csrc/distributed/rpc/tensorpipe_agent.cpp

```diff
@@ -408,35 +444,45 @@ void TensorPipeAgent::onListenerAccepted(
 
 void TensorPipeAgent::pipeRead(
     const std::shared_ptr<tensorpipe::Pipe>& pipe,
-    std::function<void(const tensorpipe::Error&, Message&&)> fn) noexcept {
-  pipe->readDescriptor([fn{std::move(fn)}, pipe](
+    std::function<void(
+        const tensorpipe::Error&,
+        Message&&,
+        std::shared_ptr<FullDeviceContext>)> fn) noexcept {
+  pipe->readDescriptor([fn{std::move(fn)}, pipe, this](
                            const tensorpipe::Error& error,
                            tensorpipe::Message tpMessage) mutable {
+    // TODO: pass these streams to TensorPipe when it can accept streams
+    auto ctx = createFullDeviceContext(
+        reverseDeviceMaps_.empty() && opts_.deviceMaps.empty());
```

Comment:
Shouldn't we only create a context for devices involved in the operation? It looks like we get a stream from the pool for all available devices.

Reply:
I agree with @pritamdamania87. As mentioned above, I think it would be good if we could be "lazy" and only access devices once we need them, to avoid initializing them uselessly.

I suspect one reason for getting a stream for all devices is in case the remote user function returns a tensor on a device that wasn't used by any of the arguments. In such a case, we would need to have set a current stream for that device before invoking the remote user function.

However, we've had to deal with a similar concern regarding CUDAFuture callbacks, and there it proved to be impossible to get a stream for all devices (it ended up causing a deadlock with NCCL), hence we've resorted to just get a stream for the devices used by the future's value, and declare that if the callback uses another device that's undefined behavior.

Hence, I think we could do the same thing for remote users functions (it would be consistent!), and maybe we can find a better solution that covers both use cases in a later release? (As a general API approach, I'd rather start with more restrictions and gradually lift them, than start with a generic but problematic solution that cannot be changed without breaking backwards compatibility).

Reply:
> hence we've resorted to just get a stream for the devices used by the future's value, and declare that if the callback uses another device that's undefined behavior.

I am not sure if this OK for RPC user functions. Suppose we are running pipeline parallel on a 8GPU machine with 2 processes. The 1st process would take input on `cuda:0` and produce output on `cuda:3`, and then send it to `cuda:4` on the other process. If we only use devices listed in the input args, does it mean users will have to move the intermediate output to `cuda:0` before sending it?

Reply:
> As mentioned above, I think it would be good if we could be "lazy" and only access devices once we need them, to avoid initializing them uselessly.

Yep, let me make this change first.

Reply:
> The 1st process would take input on cuda:0 and produce output on cuda:3

Are we sure this is a common use-case? Correct me if I'm wrong, but I believe that all CUDA kernels (i.e., the computations themselves) operate with inputs and outputs that are on the same device. Hence if a function returns an output on a different device than its inputs, it must be because the function has manually explicitly performed such a transfer, for example by calling `.to(idx)`. I could try to argue that this should be considered non-idiomatic for RPC, since RPC should be used for all data transfers, in order to provide consistency, performance and "scalability" (as in resiliency to topology changes, since it allows to easily switch from local GPUs to remote GPUs without code changes). Hence such a function should probably be split into two parts (each one operating on a single device) which should be connected through RPC calls.

Admittedly the code to do this might be a bit more convoluted than just using `.to(...)`, but I think we can always provide helpers and examples to make this easier to use. (For example, I believe #48790 might already help).

Reply:
> Hence such a function should probably be split into two parts (each one operating on a single device) which should be connected through RPC calls.

This would mean we need an overhaul to our `torch.distributed.pipe` implementation. I am open to either revamp the existing one, or implement a new distributed pipe where each process only exclusively uses one device.

Reply:
Could you fill me in on what `torch.distributed.pipe` does? Is it calling `.to()` explicitly?